web/modules/custom/nidirect_webforms/src/Plugin/WebformHandler/PrisonVisitBookingHandler.php

```diff
@@ -800,28 +763,54 @@ private function getAvailableSlots() {
   }
 
   /**
-   * Take an array of config visit slots and update the slot dates.
+   * Get data stored in cache or from file.
    */
-  private function updateConfigVisitSlotDates(array $visit_slots) {
-    $updated_slots = [];
-    $date = clone $this->bookingReference['date_valid_from'];
-    $validity_weeks = $this->bookingReference['validity_period_days'] / 7;
+  private function getData() {
+
+    $data = [];
+
+    // Face-to-face slots are retrieved from external data in cache
+    // (see PrisonVisitBookingJsonApiController.php). If there is no
+    // cached data, fallback to using slots from file.
+
+    $cached_data = \Drupal::cache()->get('prison_visit_slots_data');
+
+    if (!empty($cached_data)) {
 
-    for ($i = 0; $i < $validity_weeks; $i++) {
-      if ($i > 0) {
-        $date->modify('+1 week');
+      // There is cached data.
+      $data = $cached_data->data;
+      $cached_data_timestamp = (int) $cached_data->created;
+
+      // Check when created.
+      $now = new \DateTime('now');
+      $data_last_updated = new \DateTime('now');
+      $data_last_updated->setTimestamp($cached_data_timestamp);
+
+      // Log a warning when cached data more than 24 hours old.
+      if ($now->diff($data_last_updated)->h >= 24) {
+        $this->getLogger('prison_visits')->warning('prison_visit_slots_data cache data has not been updated in last 24 hours.');
       }
 
-      foreach ($visit_slots as $slot) {
-        $slot_p = date_parse($slot);
-        $slot_date_adjusted = new \DateTime($slot);
-        $slot_date_adjusted->setISODate($date->format('Y'), $date->format('W'), $slot_p['day']);
-        $slot_date_adjusted->setTime($slot_p['hour'], $slot_p['minute'], $slot_p['second']);
-        $updated_slots[] = $slot_date_adjusted;
+    }
+    else {
+
+      // No data in cache, so try file instead. Every time the
+      // external service posts data to prison visits api controller,
+      // data is stored in cache and written to file.
+
+      $file_uri = 'private://nidirect_webforms/prison_visit_slots_data.json';
+      $file_contents = @file_get_contents($file_uri);
```

Comment:
Using @ is not recommended. I'd use file_exists to check the file and then read the contents.